janus_server/src/binary_utils/job_driver.rs

```diff
@@ -0,0 +1,284 @@
+            // underestimate the number of jobs we can acquire, and underestimation is acceptable
+            // (we'll pick up any additional jobs on the next iteration of this loop). We can't
+            // overestimate since this task is the only place that permits are acquired.
+
+            // TODO(brandon): only acquire jobs whose batch units have not already been collected (probably by modifying acquire_incomplete_aggregation_jobs)
+            let max_acquire_count = sem.available_permits();
+            info!(max_acquire_count, "Acquiring jobs");
+            let acquired_jobs = (self.incomplete_job_acquirer)(
+                self.datastore.clone(),
+                self.worker_lease_duration,
+                max_acquire_count,
+            )
+            .await;
+            let acquired_jobs = match acquired_jobs {
+                Ok(acquired_jobs) => acquired_jobs,
+                Err(err) => {
+                    error!(?err, "Couldn't acquire jobs");
+                    // Go ahead and step job discovery delay in this error case to ensure we don't
+                    // tightly loop running transactions that will fail without any delay.
+                    job_discovery_delay = self.step_job_discovery_delay(job_discovery_delay);
+                    continue;
+                }
+            };
+            if acquired_jobs.is_empty() {
+                debug!("No jobs available");
+                job_discovery_delay = self.step_job_discovery_delay(job_discovery_delay);
+                continue;
+            }
+            info!(acquired_job_count = acquired_jobs.len(), "Acquired jobs");
+
+            // Start up tasks for each acquired job.
+            job_discovery_delay = Duration::ZERO;
+            for (task_id, vdaf, job_id, lease_expiry) in acquired_jobs {
+                task::spawn({
+                    // We acquire a semaphore in the job-discovery task rather than inside the new
+                    // job-stepper task to ensure that acquiring a permit does not race with
+                    // checking how many permits we have available in the next iteration of this
+                    // loop, to maintain the invariant that this task is the only place we acquire
+                    // permits.
+                    //
+                    // Unwrap safety: we have seen that at least `acquired_jobs.len()` permits are
+                    // available, and this task is the only task that acquires permits.
+                    let permit = Arc::clone(&sem).try_acquire_owned().unwrap();
+                    let this = Arc::clone(&self);
+                    async move {
+                        info!(?lease_expiry, "Stepping job");
+                        match time::timeout(
+                            this.effective_lease_duration(lease_expiry),
+                            (this.job_stepper)(this.datastore.clone(), vdaf, task_id, job_id),
```

Comment:
I think what you propose is _possible_, but I'm not sure how to achieve it. The difficulty is convincing the compiler that the nested async blocks and function calls will complete before the `JobDriver` value gets dropped. I think that's doable if we put the right lifetimes in the trait bounds on the closures and the futures they return, but then it's tricky to convince `tokio::task::spawn` that the closure we pass to it has lifetime `'static` (cloning `Arc` is a very easy way to meet that last requirement). I gave myself about an hour to try to get this to work, then I actually tried for two and didn't get there, and now I think I really need to move on to more important problems.

Reply:
Ah, I didn't intend to nerd-snipe you :-) it's not a big deal if it's hard to make Rust do the right thing, here -- I suppose we're at most saving an `Arc` clone.

Reply:
I spent more time on this and I think I got most of the way to a solution. Basically, we want something like:

```
struct JobDriver {
    datastore: Datastore<C>, // owned value
}
```

Reply:
Right. But in practice you have `this = Arc::clone(&self)` being moved into the spawned task already; can you use `&this.datastore` to avoid doubling up on `Arc`s? That is effectively what the previous implementation was doing (well, it passed an entire `&AggregationJobDriver`, and could freely access `AggregationJobDriver::datastore` from there).

This appears to work based on modifying the existing aggregation job driver code, and I don't see a reason why it wouldn't work here too.